File: stayops/lib/services/auth_service.dart
import 'package:http/http.dart' as http;
import 'dart:convert';
import 'package:stayops/config/api_config.dart';

class AuthService {
  // Update base URL to use the specified IP address
  static const String _baseUrl = 'http://10.203.254.101:8080/api';

  // Simplified network diagnostics method (web-compatible)
  static Future<Map<String, dynamic>> diagnoseConnection() async {
    Map<String, dynamic> diagnostics = {
      'api_responding': false,
      'errors': <String>[],
    };

    try {
      // Simple HTTP health check (removed Socket.connect - not web compatible)
      print('Testing HTTP response...');
      final response = await http.get(
        Uri.parse('http://10.203.254.101:8080/'),
        headers: {'Accept': 'application/json'},
      ).timeout(Duration(seconds: 5));

      print('Health check response: ${response.statusCode}');
      diagnostics['api_responding'] = true;
    } catch (e) {
      diagnostics['errors'].add('HTTP test failed - $e');
      print('✗ HTTP test failed: $e');
    }

    return diagnostics;
  }

  static Future<Map<String, dynamic>?> login(
      String email, String password) async {
    try {
      print('Attempting to login with email: $email');
      print('API URL: ${_baseUrl}/v1/auth/login');

      // Skip diagnostics to avoid blocking login attempts
      // final diagnostics = await diagnoseConnection();
      // if (!diagnostics['api_responding']) {
      //   print('Connection diagnostics failed. Cannot reach backend.');
      //   return null;
      // }

      final response = await http
          .post(
            Uri.parse('${_baseUrl}/v1/auth/login'),
            headers: {
              'Content-Type': 'application/json',
              'Accept': 'application/json',
              'User-Agent': 'StayOps-Flutter-App',
            },
            body: jsonEncode({
              'email': email,
              'password': password,
            }),
          )
          .timeout(Duration(seconds: 10));

      print('Login response status: ${response.statusCode}');
      print('Login response headers: ${response.headers}');
      print('Login response body: ${response.body}');

      if (response.statusCode == 200) {
        final responseData = jsonDecode(response.body);
        print('Login successful: ${responseData['guest']['fullName']}');
        return responseData;
      } else {
        print('Login failed with status: ${response.statusCode}');
        print('Error response: ${response.body}');
        return null;
      }
    } catch (e) {
      print('Login error: $e');

      // Provide specific error messages for common issues
      if (e.toString().contains('Connection refused') ||
          e.toString().contains('ERR_CONNECTION_TIMED_OUT')) {
        print(
            'Backend server is not running or not accessible at 192.168.1.100:8080');
      } else if (e.toString().contains('TimeoutException')) {
        print('Request timed out - backend may be slow or unreachable');
      } else if (e.toString().contains('Failed to fetch')) {
        print(
            'Network connectivity issue - check if backend is running and accessible');
      }

      return null;
    }
  }

  static Future<bool> testConnection() async {
    try {
      print('Testing connection to ${_baseUrl}');

      // Test the actual login endpoint with a simple request
      final response = await http.post(
        Uri.parse('${_baseUrl}/v1/auth/login'),
        headers: {
          'Content-Type': 'application/json',
          'Accept': 'application/json',
        },
      ).timeout(Duration(seconds: 5));

      print('Connection test response: ${response.statusCode}');
      print('Connection test body: ${response.body}');

      return true;
    } catch (e) {
      print('Connection test failed: $e');
      if (e.toString().contains('Connection refused') ||
          e.toString().contains('ERR_CONNECTION_TIMED_OUT')) {
        print(
            'Backend server is not running or not accessible at 10.203.254.101:8080');
      }
      return false;
    }
  }

  static Future<Map<String, dynamic>?> fetchGuestData(String guestId) async {
    try {
      print('Fetching guest data for ID: $guestId');

      // Use the full API path with v1
      final url = 'http://10.203.254.101:8080/api/v1/guests/$guestId';
      print('Guest data URL: $url'); // Add this debug line to verify URL

      final response = await http.get(
        Uri.parse(url),
        headers: {
          'Content-Type': 'application/json',
          'Accept': 'application/json',
          'User-Agent': 'StayOps-Flutter-App',
        },
      ).timeout(Duration(seconds: 10));

      print('Guest data response status: ${response.statusCode}');
      print('Guest data response headers: ${response.headers}');
      print('Guest data response body: ${response.body}');

      if (response.statusCode == 200) {
        final responseData = jsonDecode(response.body);
        print('Guest data fetched successfully');
        return responseData;
      } else if (response.statusCode == 404) {
        print('Guest not found with ID: $guestId');
        return null;
      } else {
        print('Failed to fetch guest data. Status: ${response.statusCode}');
        print('Error details: ${response.body}');
        return null;
      }
    } catch (e) {
      print('Guest data fetch error: $e');

      if (e.toString().contains('Connection refused') ||
          e.toString().contains('ERR_CONNECTION_TIMED_OUT')) {
        print(
            'Backend server is not running or not accessible at 10.203.254.101:8080');
      } else if (e.toString().contains('TimeoutException')) {
        print('Request timed out - backend may be slow or unreachable');
      } else if (e.toString().contains('Failed to fetch')) {
        print(
            'Network connectivity issue - check if backend is running and accessible');
      }

      return null;
    }
  }
}

File: stayops/src/services/authService.js
// src/services/authService.js
import { API_BASE_URL, API_ENDPOINTS, ROLE_ROUTES } from '../config/api';

class AuthService {
  /**
   * Login user with backend API
   */
  async login(email, password) {
    try {
      const response = await fetch(`${API_BASE_URL}${API_ENDPOINTS.WEB_LOGIN}`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'ngrok-skip-browser-warning': 'true',
        },
        body: JSON.stringify({ email, password })
      });

      if (!response.ok) {
        const errorData = await response.json().catch(() => ({}));
        throw new Error(errorData.error || 'Invalid email or password');
      }

      const data = await response.json();
      
      // Transform backend response to frontend format
      const userData = {
        id: data.userId,
        email: data.email,
        name: data.fullName || data.username,
        role: data.role,
        username: data.username,
        portal: ROLE_ROUTES[data.role] || '/dashboard',
        token: data.token || `token_${Date.now()}`, // Use backend token or generate temporary one
        entityId: data.entityId
      };

      return userData;
    } catch (error) {
      console.error('Login error:', error);
      throw error;
    }
  }

  /**
   * Validate token with backend
   */
  async validateToken(token) {
    try {
      const response = await fetch(
        `${API_BASE_URL}${API_ENDPOINTS.VALIDATE_TOKEN}?token=${token}`,
        {
          method: 'GET',
          headers: {
            'Authorization': `Bearer ${token}`,
            'ngrok-skip-browser-warning': 'true',
          }
        }
      );

      if (!response.ok) {
        return false;
      }

      const data = await response.json();
      return data.valid || false;
    } catch (error) {
      console.error('Token validation error:', error);
      return false;
    }
  }

  /**
   * Logout user
   */
  async logout() {
    try {
      await fetch(`${API_BASE_URL}${API_ENDPOINTS.LOGOUT}`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'ngrok-skip-browser-warning': 'true',
        }
      });
      return true;
    } catch (error) {
      console.error('Logout error:', error);
      return false;
    }
  }

  /**
   * Debug: Get all users from backend
   */
  async debugUsers() {
    try {
      const response = await fetch(`${API_BASE_URL}${API_ENDPOINTS.DEBUG_USERS}`);
      const data = await response.json();
      return data;
    } catch (error) {
      console.error('Debug users error:', error);
      return null;
    }
  }

  /**
   * Get user by email
   */
  async getUserByEmail(email) {
    try {
      const response = await fetch(`${API_BASE_URL}${API_ENDPOINTS.USER_BY_EMAIL(email)}`);
      if (!response.ok) {
        throw new Error('User not found');
      }
      return await response.json();
    } catch (error) {
      console.error('Get user error:', error);
      throw error;
    }
  }
}

export default new AuthService();